data_steward/cdr_cleaner/cleaning_rules/update_fields_numbers_as_strings.py

```diff
@@ -0,0 +1,175 @@
+"""
+Original Issues: DC-1052
+
+Background
+It has been discovered that the field type for some PPI survey answers is incorrect: there are several instances of
+numeric answers being saved as ‘string’ field types. The expected long-term fix is for PTSC to correct the field type
+on their end; however, there is no anticipated timeline for the completion of this work. As a result, the Curation team
+will need to create a cleaning rule to correct these errors.
+
+This cleaning rule should only apply to specific numeric values collected in COPE surveys, a list of specific
+observation_source_values that are affected in code.
+
+Cleaning rule to fill null values in value_as_number with values in value_as_string,
+EXCEPT when it’s a ‘PMI Skip’ for each of the observation_source_value
+
+Rule should be applied to the RDR export
+"""
+
+# Project imports
+import constants.cdr_cleaner.clean_cdr as cdr_consts
+from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
+from constants.bq_utils import WRITE_TRUNCATE
+from utils.bq import JINJA_ENV
+
+# observation_source_values to replace
+OBSERVATION_SOURCE_VALUES = "('basics_xx', 'basics_xx20', 'ipaq_1_cope_a_24', 'ipaq_2_cope_a_160', " \
+                            "'ipaq_2_cope_a_85', 'ipaq_3_cope_a_24', 'ipaq_4_cope_a_160', 'ipaq_4_cope_a_85', " \
+                            "'ipaq_5_cope_a_24', 'ipaq_6_cope_a_160', 'ipaq_6_cope_a_85', 'cope_a_160', 'cope_a_85', " \
+                            "'copect_50_xx19_cope_a_57', 'copect_50_xx19_cope_a_198', 'copect_50_xx19_cope_a_152', " \
+                            "'lifestyle_2_xx12_cope_a_57', 'lifestyle_2_xx12_cope_a_198', " \
+                            "'lifestyle_2_xx12_cope_a_152', 'ipaq_1_cope_a_85', 'ipaq_3_cope_a_85', " \
+                            "'ipaq_5_cope_a_85', 'ipaq_7_cope_a_160', 'ipaq_7_cope_a_85', 'cdc_covid_19_n_a2')"
+
+# Query to sandbox original observation table before CR
+SANDBOX_QUERY = JINJA_ENV.from_string("""
+CREATE OR REPLACE TABLE `{{project}}.{{sandbox_dataset}}.{{sandbox_table}}` as(
+SELECT * FROM
+`{{project}}.{{dataset}}.observation`
+)
+""")
+
+# Query to update value_as_number field
+NUMBERS_AS_STRINGS_QUERY = JINJA_ENV.from_string("""
+SELECT
+  observation_id,
+  person_id,
+  observation_concept_id,
+  observation_date,
+  observation_datetime,
+  observation_type_concept_id,
+  CASE WHEN observation_source_value IN {{observation_source_values}}
+  AND value_as_number IS NULL AND value_as_string != 'PMI Skip' AND REGEXP_CONTAINS(value_as_string, r'^\d+$')
```

Comment:
can you explain `REGEXP_CONTAINS(value_as_string, r'^\d+$'`

Reply:
This is using regex to make sure the string in that field only contains numbers, nothing else.